### ISSUE TITLE
mol2 writer changes bond orders !!

### ISSUE THREAD

**Issue by davidlmobley** [ParmEd/ParmEd#691]
The `.mol2` writer seems to be changing bond orders, which is bad news for us (we need to retain bond orders). Hopefully there's a simple fix? For example, see:

```
import parmed

name = 'toluene'

mol2_input_filename = name+'_noparmed.mol2'
mol2_output_filename = name + '.mol2'

mol2f = parmed.formats.Mol2File
mol2f.write(parmed.load_file(mol2_input_filename).fix_charges(),mol2_output_filename)
```

using the attached toluene_noparmed.mol2 as input. 

For example, see the original `bonds` section:
```
@<TRIPOS>BOND
     1    1    2 1
     2    2    7 ar
     3    2    3 ar
     4    3    4 ar
     5    4    5 ar
     6    5    6 ar
     7    6    7 ar
     8    1    8 1
     9    1    9 1
    10    1   10 1
    11    3   11 1
    12    4   12 1
    13    5   13 1
    14    6   14 1
    15    7   15 1
```

and the new one:
```
@<TRIPOS>BOND
       1        1        2 1
       2        2        7 1
       3        2        3 1
       4        3        4 1
       5        4        5 1
       6        5        6 1
       7        6        7 1
       8        1        8 1
       9        1        9 1
      10        1       10 1
      11        3       11 1
      12        4       12 1
      13        5       13 1
      14        6       14 1
      15        7       15 1
```

(Thanks, @camizanette )

[toluene.mol2.zip](https://github.com/ParmEd/ParmEd/files/224827/toluene.mol2.zip)

**Comment by davidlmobley** [comment]
Er, OK, so looking through the code I'm thinking the situation is actually not good in terms of resolving this. Specifically, it looks like you may ignore the bond orders entirely, @swails ? In which case, writing a `.mol2` file would ALWAYS result in single bonds regardless of what bond orders were present initially? If so, then you'll need to add BIG BIG disclaimers to your `.mol2` writer, as any `.mol2` files written by ParmEd would then result in big downstream problems if they were processed again by prep tools (antechamber, OpenEye toolkits, etc.).

**Comment by davidlmobley** [comment]
... that is, IF I'm reading it right. :) 

Longer-term, it would probably be VERY VERY helpful to have this fixed - not just in the writer, but the ability to keep track of bond orders internally. One direction we are likely headed for some future forcefield efforts is to start paying more attention to bond orders within molecular simulations and having these play a role in parameter assignment. That will mean we'll want to hold on to bond order info.

**Comment by hainm** [comment]
```python
     2    2    7 ar
     3    2    3 ar
     4    3    4 ar
     5    4    5 ar
     6    5    6 ar
     7    6    7 ar
```

```python
       2        2        7 1
       3        2        3 1
       4        3        4 1
       5        4        5 1
       6        5        6 1
       7        6        7 1
```

@davidlmobley  what does 'ar' mean?

**Comment by davidlmobley** [comment]
@hainm - aromatic

**Comment by hainm** [comment]
so what you meant is replacing '2    2    7  1' by '2    2    7 ar', right?

**Comment by davidlmobley** [comment]
Yup, the code just writes the value `1` for every bond type. Ugh:
```
https://github.com/ParmEd/ParmEd/blob/master/parmed/formats/mol2.py#L469
```

I realize that forcefields don't currently care about bond type, but `.mol2` files sure do. I'd hesitate to call these real `.mol2` files if the bond type/order is not preserved.

**Comment by davidlmobley** [comment]
@hainm 
>so what you meant is replacing '2 2 7 1' by '2 2 7 ar', right?

It's replacing the correct `2 2 7 ar` with `2 2 7 1`. Worse, it would replace `N1 N2 N3 3` with `N1 N2 N3 1` thereby changing a triple bond into a single bond.

**Comment by davidlmobley** [comment]
I'll have to get my group to think through whether we've done anything with ParmEd that this would have broken. I don't THINK we've written to the `.mol2` format much yet, but we were using `fix_charges` (https://github.com/ParmEd/ParmEd/blob/9b187500363e69c441ef0e356d601899ce7bfde2/parmed/modeller/residue.py#L264) to make sure residues came out to integer charge, and if we do *any* downstream processing of the files after running this on them (via antechamber or OpenEye tools, RDKit, etc.) this would break it.  The one case I'm looking at so far we go straight into `tleap` after using the `.mol2` writer so it doesn't seem to have broken anything, but there may be others I haven't remembered yet.

**Comment by ctk3b** [comment]
As far as I can tell, bond orders are never even read in: https://github.com/ParmEd/ParmEd/blob/master/parmed/formats/mol2.py#L196

**Comment by davidlmobley** [comment]
Yep.

This seems not to have caused us to run any incorrect simulations *yet* (because we've aways used ParmEd only *after* assigning atom types via antechamber), but it will certainly lead to incorrect simulations down the line.

I'll open issues on openmoltools and our SolvationToolkit to bypass this code until there is a fix.

**Comment by swails** [comment]
I've never personally dealt with a scenario where bond orders were used, which is why the mol2 writer behaves the way it does (all tools in the Amber set the bond order to 1 and ignore it when reading as far as I remember).  Since no other file formats supported by ParmEd contain this information, retaining it served no purpose in file format conversions.

Which brings me to my question: what is your use-case here?  It would be easy to have ParmEd simply remember the bond orders and print them back out when going from mol2 to mol2.  Bond type perception will be a lot bigger project (but something that will be much easier after #668).

mol2->mol2 remembering bond order will be done tonight.  Is that enough for your uses?

**Comment by davidlmobley** [comment]
> mol2->mol2 remembering bond order will be done tonight. Is that enough for your uses?

Yes! That's enough for now. Our current problem is just that we set things up for simulations, do some manipulation of final `.mol2` files via ParmEd, and then proceed to simulations. We expect the `.mol2` files to have a record of what molecules we simulated, and no one downstream would remember that the bond orders got messed up. So as long as our `.mol2` to `.mol2` manipulations in ParmEd don't perturb bond orders we are OK for now.

> I've never personally dealt with a scenario where bond orders were used, which is why the mol2 writer behaves the way it does (all tools in the Amber set the bond order to 1 and ignore it when reading as far as I remember). Since no other file formats supported by ParmEd contain this information, retaining it served no purpose in file format conversions.

Yes, this makes sense. Normally forcefields have no notion of bond order. My, er, horror here was mostly in finding that a file format which DOES have a notion of bond order loses such notions. 

>Which brings me to my question: what is your use-case here? It would be easy to have ParmEd simply remember the bond orders and print them back out when going from mol2 to mol2. Bond type perception will be a lot bigger project (but something that will be much easier after #668).

Ultimate use-case: Building a new forcefield which has bond types which *are* remembered and used for determining parameters - i.e. one could develop bond parameters for bond orders 1 and 2 and if a given environment means a bond is somewhere in between bond order 1 and 2, it could have a bond type which is intermediate (yes, non-integer) and this would be used to determine parameters (this would be decoupled from our ideas of "atom type" - i.e. atoms could retain the same type as far as sigma and epsilon, but their bond could change "order" depending on environment). This isn't your job, of course. I just thought I'd mention it since retaining bond orders is sort of a related point -- that is, just because they don't mean anything for forcefields now doesn't mean they always will.

**Comment by swails** [comment]
OK.  This is implemented by #692.  I'll tag a release after this gets merged in.  In any case, versioneer is tagging that commit as `ParmEd==2.5.0+31.g6c85623`.  So if you want to ensure your scripts are running with a version of ParmEd that retains bond order, you can add something like this:

```Python
import parmed as pmd

if pmd.version < (2, 5, 0, 31):
    raise ImportError('Insufficient ParmEd version... update!')
```

Versioneer is pretty convenient so far.  Just make sure that you pull all tags from the main repo into your repository (or use pip or conda after I upload a new version of the package).

**Comment by davidlmobley** [comment]
Thanks, that was super fast!

David


On Mon, Apr 18, 2016 at 6:50 PM, Jason Swails <notifications@github.com>
wrote:

> OK. This is implemented by #692
> <https://github.com/ParmEd/ParmEd/pull/692>. I'll tag a release after
> this gets merged in. In any case, versioneer is tagging that commit as
> ParmEd==2.5.0+31.g6c85623. So if you want to ensure your scripts are
> running with a version of ParmEd that retains bond order, you can add
> something like this:
>
> import parmed as pmd
> if pmd.version < (2, 5, 0, 31):
>     raise ImportError('Insufficient ParmEd version... update!')
>
> Versioneer is pretty convenient so far. Just make sure that you pull all
> tags from the main repo into your repository (or use pip or conda after I
> upload a new version of the package).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/691#issuecomment-211679942>
>



-- 
David Mobley
dmobley@gmail.com
949-385-2436
